test/test_forbidden_sequence.jl

```diff
@@ -0,0 +1,305 @@
+@testset verbose=false "Forbidden Sequence" begin
+    function dummy_tree(sequence)::AbstractRuleNode
+        # returns a tree that contains the specified sequence and some noise.
+        # holes can be represented by tuples of indices
+        if length(sequence) == 1
+            if sequence[1] isa Tuple
+                domain = BitVector((r ∈ sequence[1] for r ∈ 1:9))
+                return Hole(domain)
+            end
+            return RuleNode(sequence[1])
+        end
+        children = [dummy_tree(sequence[2:end])]
+        if sequence[1] isa Tuple
+            domain = BitVector((r ∈ sequence[1] for r ∈ 1:9))
+            return UniformHole(domain, children)
+        end
+        return RuleNode(sequence[1], children)
+    end
+    
+    function dummy_solver(sequence, constraint)::Solver
+        grammar = @csgrammar begin
+            S = (S, 1) | (S, 2) | (S, 3) | (S, 4) | (S, 5) | (S, 6) | (S, 7) | (S, 8)
+            S = 9
+        end
+        # only 1 grammar is ever instantiated (at compile time), so we need to clear the grammar
+        empty!(grammar.constraints) 
+        addconstraint!(grammar, constraint)
+        return GenericSolver(grammar, dummy_tree(sequence))
+    end
+    
+    function get_sequence(solver, n)
+        #converts a tree into a sequence of Int (representing a rule) and Tuple{Int} (representing a domain of rules)
+        sequence = []
+        node = get_tree(solver)
+        for _ ∈ 1:n
+            if isfilled(node)
+                push!(sequence, get_rule(node))
+            else
+                push!(sequence, Tuple(rule for rule ∈ 1:9 if node.domain[rule]))
+            end
+            if !isempty(get_children(node))
+                node = get_children(node)[1]
+            end
+        end
+        return sequence
+    end
+    
+    function test_propagation(
+        constraint::ForbiddenSequence, 
+        sequence_before_propagation, 
+        sequence_after_propagation
+    )
+        # An "Int" in the sequence represents a rule (RuleNode)
+        # A "Tuple" in the sequence represents a domain of rules (UniformHole)
+        solver = dummy_solver(sequence_before_propagation, constraint)
+        actual = get_sequence(solver, length(sequence_before_propagation))
+        expected = sequence_after_propagation
+        @test actual == expected
+    end
+    
+    function test_infeasible(
+        constraint::ForbiddenSequence, 
+        sequence_before_propagation
+    )
+        solver = dummy_solver(sequence_before_propagation, constraint)
+        @test isfeasible(solver) == false
+    end
+
+    @testset "propagate" begin
+        constraint = ForbiddenSequence([1, 2, 3])
+        solver = dummy_solver([1, 2, 3], constraint)
+        tree = get_tree(solver)
+        @assert true
```

Comment:
Is this test supposed to check that no errors are thrown in the lines above? `@assert true` is confusing to see at first.